gym/vector/sync_vector_env.py

```diff
@@ -0,0 +1,113 @@
+        self.copy = copy
+        self.episodic = episodic
+
+        if (observation_space is None) or (action_space is None):
+            observation_space = observation_space or self.envs[0].observation_space
+            action_space = action_space or self.envs[0].action_space
+        super(SyncVectorEnv, self).__init__(num_envs=len(env_fns),
+            observation_space=observation_space, action_space=action_space)
+
+        self._check_observation_spaces()
+        self.observations = create_empty_array(self.single_observation_space,
+            n=self.num_envs, fn=np.empty)
+        self._rewards = np.zeros((self.num_envs,), dtype=np.float64)
+        self._dones = np.zeros((self.num_envs,), dtype=np.bool_)
+        self._zero_observation = create_empty_array(self.single_observation_space,
+            n=None, fn=np.zeros)
+
+    def reset(self):
+        self._dones[:] = False
+        observations = []
+        for i in range(self.num_envs):
+            observation = self.envs[i].reset()
+            observations.append(observation)
+        concatenate(observations, self.observations, self.single_observation_space)
+
+        return np.copy(self.observations) if self.copy else self.observations
+
+    def step(self, actions):
+        observations, infos = [], []
+        for i, action in enumerate(actions):
+
+            if self.episodic and self._dones[i]:
+                observations.append(self._zero_observation)
+                self._rewards[i] = 0.
+                infos.append({})
+                continue
+
+            observation, self._rewards[i], self._dones[i], info = self.envs[i].step(action)
+            if self._dones[i]:
+                if self.episodic:
+                    observation = self._zero_observation
+                    info.update({'SyncVectorEnv.end_episode': True})
+                else:
+                    observation =  self.env[i].reset()
+
+            observations.append(observation)
+            infos.append(info)
+        concatenate(observations, self.observations, self.single_observation_space)
+
+        return (np.copy(self.observations) if self.copy else self.observations,
+            np.copy(self._rewards), np.copy(self._dones), infos)
+
+    def close_extras(self):
+        for env in self.envs:
+            env.close()
+
+    def _check_observation_spaces(self):
+        from gym.spaces import Box
+        for env in self.envs:
+            observation_space = env.observation_space
+            # Equality between Box spaces does not check for shape equality
+            if isinstance(observation_space, Box) \
+                    and isinstance(self.single_observation_space, Box) \
+                    and observation_space.shape != self.single_observation_space.shape:
+                break
+            if not (observation_space == self.single_observation_space):
+                break
+        else:
+            return True
+        self.close()
```

Comment:
Seems like the user should catch exceptions and close environments rather than them automatically closing themselves.